File: VB/ColumnSpacing/MyGridPainter.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports DevExpress.XtraGrid.Views.Grid.Drawing
Imports DevExpress.XtraGrid.Views.Grid
Imports System.Drawing
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo
Imports DevExpress.XtraGrid.Drawing
Imports DevExpress.XtraGrid.Views.Base

Namespace ColumnSpacing
	Friend Class MyGridPainter
		Inherits GridPainter
		Public Sub New(ByVal view As GridView)
			MyBase.New(view)
		End Sub

		Protected Overrides Sub DrawContents(ByVal e As GridViewDrawArgs)
			MyBase.DrawContents(e)
			DrawSpacingColumns(e)
		End Sub


		Protected Overridable Sub DrawSpacingColumns(ByVal e As GridViewDrawArgs)
			Dim columnBounds As RectangleF
			Dim columnSpacing As RectangleF

			For i As Integer = 0 To View.VisibleColumns.Count - 1
				Dim myColumn = TryCast(View.VisibleColumns(i), MyGridColumn)
				columnBounds = GetColumnBounds(TryCast(View, MyGridView), myColumn)
                If columnBounds.IsEmpty Then
                    columnBounds.Height -= 2
                    If i > 0 Then
                        columnSpacing = New RectangleF(columnBounds.Left - myColumn.ColumnSpacing, columnBounds.Top, myColumn.ColumnSpacing, columnBounds.Height)
                        e.Graphics.FillRectangle(e.ViewInfo.PaintAppearance.Empty.GetBackBrush(e.Cache), columnSpacing)
                    End If

                    If (i = View.VisibleColumns.Count - 1) AndAlso (myColumn.ColumnSpacing <> 0) Then
                        columnSpacing = New RectangleF(columnBounds.Left + columnBounds.Width, columnBounds.Top, View.ViewRect.Width, columnBounds.Height)
                    Else
                        columnSpacing = New RectangleF(columnBounds.Left + columnBounds.Width, columnBounds.Top, myColumn.ColumnSpacing, columnBounds.Height)
                    End If
                    e.Graphics.FillRectangle(e.ViewInfo.PaintAppearance.Empty.GetBackBrush(e.Cache), columnSpacing)
                End If

                columnBounds.X = columnBounds.Left - 1
                columnBounds.Width = 1
                columnBounds.Height += 1
                e.Graphics.FillRectangle(e.ViewInfo.PaintAppearance.VertLine.GetBackBrush(e.Cache), columnBounds)
            Next i
		End Sub

		Private Function GetColumnHeaderBounds(ByVal view As MyGridView, ByVal col As MyGridColumn) As RectangleF
			Dim viewInfo As GridViewInfo = TryCast(view.GetViewInfo(), GridViewInfo)
			If viewInfo.IsReady AndAlso (Not viewInfo.IsDataDirty) Then
				For Each info As GridColumnInfoArgs In viewInfo.ColumnsInfo
					If info.Column Is col Then
						Return info.Bounds
					End If
				Next info
			End If
			Return RectangleF.Empty
		End Function

		Private Function GetColumnBounds(ByVal view As MyGridView, ByVal col As MyGridColumn) As RectangleF
			Dim result As RectangleF = GetColumnHeaderBounds(view, col)
			If result.IsEmpty Then
				Return result
			End If
			Dim viewInfo As GridViewInfo = TryCast(view.GetViewInfo(), GridViewInfo)
			Dim rowsHeight As Integer = viewInfo.RowsLoadInfo.ResultRowsHeight
			result.Height += rowsHeight
			Return result
		End Function
	End Class
End Namespace

File: CS/ColumnSpacing/MyGridPainter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using DevExpress.XtraGrid.Views.Grid.Drawing;
using DevExpress.XtraGrid.Views.Grid;
using System.Drawing;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using DevExpress.XtraGrid.Drawing;
using DevExpress.XtraGrid.Views.Base;

namespace ColumnSpacing
{
    class MyGridPainter : GridPainter
    {
        public MyGridPainter(GridView view) : base(view) {}

        protected override void DrawContents(GridViewDrawArgs e)
        {
            base.DrawContents(e);
            DrawSpacingColumns(e);
        }


        protected virtual void DrawSpacingColumns(GridViewDrawArgs e)
        {
            RectangleF columnBounds;
            RectangleF columnSpacing;

            for (int i = 0; i < View.VisibleColumns.Count; i++)
            {
                var myColumn = View.VisibleColumns[i] as MyGridColumn;
                columnBounds = GetColumnBounds(View as MyGridView, myColumn);
                if (columnBounds != null)
                {
                    columnBounds.Height -= 2;
                    if (i > 0)
                    {
                        columnSpacing = new RectangleF(columnBounds.Left - myColumn.ColumnSpacing, columnBounds.Top, myColumn.ColumnSpacing, columnBounds.Height);
                        e.Graphics.FillRectangle(e.ViewInfo.PaintAppearance.Empty.GetBackBrush(e.Cache), columnSpacing);
                    }

                    if ((i == View.VisibleColumns.Count - 1) && (myColumn.ColumnSpacing != 0))
                    {
                        columnSpacing = new RectangleF(columnBounds.Left + columnBounds.Width, columnBounds.Top, View.ViewRect.Width, columnBounds.Height);
                    }
                    else
                    {
                        columnSpacing = new RectangleF(columnBounds.Left + columnBounds.Width, columnBounds.Top, myColumn.ColumnSpacing, columnBounds.Height);
                    }
                    e.Graphics.FillRectangle(e.ViewInfo.PaintAppearance.Empty.GetBackBrush(e.Cache), columnSpacing);

                    columnBounds.X = columnBounds.Left - 1;
                    columnBounds.Width = 1;
                    columnBounds.Height += 1;

                    e.Graphics.FillRectangle(e.ViewInfo.PaintAppearance.VertLine.GetBackBrush(e.Cache), columnBounds);
                }
            }
        }

        RectangleF GetColumnHeaderBounds(MyGridView view, MyGridColumn col)
        {
            GridViewInfo viewInfo = view.GetViewInfo() as GridViewInfo;
            if (viewInfo.IsReady && !viewInfo.IsDataDirty)
            {
                foreach (GridColumnInfoArgs info in viewInfo.ColumnsInfo)
                {
                    if (info.Column == col)
                        return info.Bounds;
                }
            }
            return RectangleF.Empty;
        }

        RectangleF GetColumnBounds(MyGridView view, MyGridColumn col)
        {
            RectangleF result = GetColumnHeaderBounds(view, col);
            if (result.IsEmpty)
                return result;
            GridViewInfo viewInfo = view.GetViewInfo() as GridViewInfo;
            int rowsHeight = viewInfo.RowsLoadInfo.ResultRowsHeight;
            result.Height += rowsHeight;
            return result;
        }
    }
}
